pairing/plonk/src/prover.rs

```diff
@@ -0,0 +1,293 @@
+
+        // compute public input polynomial
+        let public = witness[0..index.cs.public].to_vec();
+        let p = -Evaluations::<E::Fr>::from_vec_and_domain(public.clone(), index.cs.domain).interpolate();
+
+        let a = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.l.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+        let b = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.r.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+        let c = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.o.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+
+        // commit to the a, b, c wire values
+        let a_comm = index.urs.get_ref().commit(&a)?;
+        let b_comm = index.urs.get_ref().commit(&b)?;
+        let c_comm = index.urs.get_ref().commit(&c)?;
+
+        // absorb the public input, a, b, c polycommitments into the argument
+        fq_sponge.absorb_fr(&public);
+        fq_sponge.absorb_g(&[a_comm, b_comm, c_comm]);
+
+        // sample beta, gamma oracles
+        oracles.beta = fq_sponge.challenge();
+        oracles.gamma = fq_sponge.challenge();
+
+        // compute permutation polynomial
+
+        let mut z = (0..n).map
+        (
+            |j|
+                (witness[j] + &(index.cs.sigmal[0][j] * &oracles.beta) + &oracles.gamma) *&
+                (witness[j+n] + &(index.cs.sigmal[1][j] * &oracles.beta) + &oracles.gamma) *&
+                (witness[j+2*n] + &(index.cs.sigmal[2][j] * &oracles.beta) + &oracles.gamma)
+        ).collect::<Vec<_>>();
+        
+        algebra::fields::batch_inversion::<E::Fr>(&mut z);
+
+        (0..n).for_each
+        (
+            |j| z[j] *=
+                &((witness[j] + &(index.cs.sid[j] * &oracles.beta) + &oracles.gamma) *&
+                (witness[j+n] + &(index.cs.sid[j] * &oracles.beta * &index.cs.r) + &oracles.gamma) *&
+                (witness[j+2*n] + &(index.cs.sid[j] * &oracles.beta * &index.cs.o) + &oracles.gamma))
+        );
+
+        (1..n).for_each(|i| {let x = z[i-1]; z[i] *= &x});
+        if z.pop().unwrap() != E::Fr::one() {return Err(ProofError::ProofCreation)};
+        z.insert(0, E::Fr::one());
+        
+        let z = Evaluations::<E::Fr>::from_vec_and_domain(z, index.cs.domain).interpolate();
+
+        // commit to z
+        let z_comm = index.urs.get_ref().commit(&z)?;
+
+        // absorb the z commitment into the argument and query alpha
+        fq_sponge.absorb_g(&[z_comm]);
+        oracles.alpha = fq_sponge.challenge();
+        let alpsq = oracles.alpha.square();
+
+        // compute quotient polynomial
+
+        let t1 =
```

Comment:
Is there some way to make this more readable? I think some of these borrows are not necessary (I believe only the RHS of all the operations needs to be a borrow).

Reply:
I tried few ways of making it less horrible. Unfortunately, + operator's add function takes both left as well as right sides as references. And, although we can remove few references from these expressions, the primary inconvenience is with expressions consisting of many arith operations which add up the references. Therefore, not referencing some of the operands will not make it much better. I may have misunderstood your suggestion though. If so, please clarify.